services/motion/builtin/builtin.go

```diff
@@ -231,52 +233,183 @@ func (ms *builtIn) MoveOnGlobe(
 
-	plan, kb, err := ms.planMoveOnGlobe(
-		ctx,
-		componentName,
-		destination,
-		movementSensorName,
-		obstacles,
-		kinematicsOptions,
-		extra,
-	)
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	movementSensor, ok := ms.movementSensors[movementSensorName]
+	if !ok {
+		return false, resource.DependencyNotFoundError(movementSensorName)
+	}
+
+	successChan := make(chan bool)
+	defer close(successChan)
+
+	// TODO: I know this has to change
+	type planAndBase struct {
+		base kinematicbase.KinematicBase
+		plan motionplan.Plan
+	}
+	planChan := make(chan planAndBase)
+	defer close(planChan)
+
+	replanChan := make(chan bool, 1)
+	defer close(replanChan)
+	replanChan <- true
+
+	errChan := make(chan error)
+	defer close(errChan)
+
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	ms.cancelFn = cancelFn
+	defer ms.cancelFn()
+
+	var backgroundWorkers sync.WaitGroup
+	defer backgroundWorkers.Wait()
+
+	// helper function to manage polling functions
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
+		backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			ticker := time.NewTicker(period)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					errChan <- ctx.Err()
+					return
+				case <-ticker.C:
+					if err := fn(ctx); err != nil {
+						errChan <- err
+					}
+					return
+				}
+			}
+		}, backgroundWorkers.Done)
+	}
+
+	// start goroutine to execute plan
+	backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		for {
+			select {
+			case <-ctx.Done():
+				errChan <- ctx.Err()
+				return
+			case planBase := <-planChan:
+				if err := ms.executePlan(cancelCtx, planBase.base, planBase.plan); err != nil {
+					errChan <- err
+					return
+				}
+
+				position, _, err := movementSensor.Position(cancelCtx, nil)
+				if err != nil {
+					errChan <- err
+					return
+				}
+
+				ms.logger.Infof("position: %v", position)
+				if spatialmath.GeoPointToPose(position, destination).Point().Norm() <= kinematicsOptions.PlanDeviationThresholdMM {
+					successChan <- true
+					return
+				}
+			}
+		}
+	}, backgroundWorkers.Done)
+
+	// loop to monitor things that poll and replan appropriately
+	for {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case err := <-errChan:
+			fmt.Println("this actually happening")
+			return false, err
+		case <-replanChan:
+			fmt.Println("replanning")
+
+			ms.cancelFn()
+			cancelCtx, cancelFn = context.WithCancel(ctx)
+			ms.cancelFn = cancelFn
+
+			plan, kb, err := ms.planMoveOnGlobe(
+				ctx,
+				componentName,
+				destination,
+				movementSensor,
+				obstacles,
+				kinematicsOptions,
+				extra,
+			)
+			if err != nil {
+				return false, err
+			}
+			planChan <- planAndBase{kb, plan}
+
+			// drain the replanChan
```

Comment:
[q] why does the `replanChan` need to explicitly drained here? I would except that here it'd always be empty.

Reply:
Yeah this is something that I thought about for a while and decided this was the safest thing to do but there might be a better way to do it.  

I have a buffered channel that has two writers (potentially more in the future) and one reader.  If both writers write we want to make sure that we don't have anything left in the channel that would immediately cause another replan